http://cygwin.com/ml/cygwin-xfree/2009-06/msg00075.html

The reason that the binding is failing is that the address to bind to is 
incorrectly set in /usr/include/X11/Xtrans/Xtranssock.c:TRANS(SocketINETCreateListener)
as the value of in6addr_any seems to be "SocketINETGetAddr" rather than the 
expected all zeroes.

I think perhaps this is a problem with gcc4's weak symbol handling, as 
if I remove the '#pragma weak in6addr_any = local_in6addr_any' from 
Xtranssock.c, things work as expected (with the pragma present, the 
address used for in6addr_any seems to be slightly offset from 
local_in6addr_any...)

--- origsrc/xtrans-1.2.3/Xtranssock.c	2008-08-07 17:11:52.000000000 -0500
+++ src/xtrans-1.2.3/Xtranssock.c	2009-06-29 19:53:04.837416800 -0500
@@ -233,8 +233,10 @@ static int TRANS(SocketINETClose) (Xtran
  * that don't have IPv6 support.
  */
 #if defined(IPv6) && defined(AF_INET6)
+#ifndef __CYGWIN__
 static const struct in6_addr local_in6addr_any = IN6ADDR_ANY_INIT;
 #pragma weak in6addr_any = local_in6addr_any
+#endif
 #ifndef __USLC__
 #pragma weak getaddrinfo
 #endif
